### PR TITLE
Do some simple constant propagation in the ConstProp pass

### PR DESCRIPTION
**Pull request by wesleywiser**
r? @oli-obk 

I added a few test cases. Let me know if you think there should be more.

### CODE REVIEW

File: src/librustc_mir/transform/const_prop.rs [hunk 1]
```diff
@@ -497,6 +497,67 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
             },
         }
     }
+
+    fn replace_with_const(&self, rval: &mut Box<Rvalue<'tcx>>, value: Const<'tcx>) {
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs]: I think it should also work to turn the `rval` argument's type into `&mut Rvalue<'tcx>` in order to save the additional heap allocation (and the dropping of the old one).

File: src/librustc_mir/transform/const_prop.rs [hunk 2]
```diff
@@ -497,6 +497,67 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
...
+        let operand = value;
+        if let interpret::Operand::Immediate(im) = *operand {
+            match im {
+                interpret::Immediate::Scalar(ScalarMaybeUndef::Scalar(scalar)) => {
+                    *rval = Box::new(Rvalue::Use(Operand::Constant(Box::new(
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs]: create a helper function for obtaining an `Operand` from a `ty::Const` (or maybe even from a `Scalar` + `Ty`, although in the future we'll probably need the `ty::Const`, too)

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_mir/transform/const_prop.rs [hunk 3]
```diff
@@ -497,6 +497,67 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
             },
         }
     }
+
+    fn replace_with_const(&self, rval: &mut Box<Rvalue<'tcx>>, value: Const<'tcx>) {
+        let operand = value;
```
Comment (oli-obk) [inline src/librustc_mir/transform/const_prop.rs]: rename the `value` argument to `operand` instead of rebinding

**Comment by wesleywiser** [comment]
Feedback resolved

**Review by oli-obk — APPROVED** [review]

**Comment by oli-obk** [comment]
@rust-timer build 7f68fd8986c182e39d7f3b125d953b55c3544568

**Comment by oli-obk** [comment]
r? @eddyb 

This is a very conservative constant propagator that will only replace the rhs of assignments with constants. Any other avenues for constant propagation are left for future changes.

**Comment by wesleywiser** [comment]
@oli-obk This is gated on `mir_opt_level >= 3` so a perf run will only measure the overhead of that check. Did you want to see the performance of running the propagation? I can push a temporary commit which un-gates the constant propagation.

**Comment by oli-obk** [comment]
> Did you want to see the performance of running the propagation? I can push a temporary commit which un-gates the constant propagation.

It's unlikely that this would cause performance changes of the compiled binary. LLVM should take care of that just fine. I wanted to run  perfbot to see whether the const propagation slows down compilation. I assumed our perf test suite runs with full MIR optimizations, but I should have checked. I'll do that now

**Comment by wesleywiser** [comment]
`mir-opt-level >= 2` results in breakage so I doubt it.

**Comment by oli-obk** [comment]
Right. So we can check the perf when we decrease the minimum level where the propagator is turned on or make O2 actually pass all tests

**Comment by wesleywiser** [comment]
Perf seems ok

File: src/test/mir-opt/const_prop/slice_len.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+//      _5 = Lt(_3, _4);
+//      assert(move _5, "index out of bounds: the len is move _4 but the index is _3") -> bb2;
+//  }
+//  bb2: {
+//      _1 = (*_2)[_3];
+//      ...
+//      return;
+//  }
+// END rustc.main.ConstProp.before.mir
+// START rustc.main.ConstProp.after.mir
```
Comment (bjorn3) [inline src/test/mir-opt/const_prop/slice_len.rs]: Mir after const prop doesn't seem to be different from before.

Comment (wesleywiser) [reply src/test/mir-opt/const_prop/slice_len.rs]: That's correct. I just added this test to try to "lock in" more of the behavior. I have additional changes to improve the constant propagation code incoming but per @oli-obk's feedback, I wanted to keep this PR small.

File: src/librustc_mir/transform/const_prop.rs [hunk 5]
```diff
@@ -497,6 +499,46 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
             },
         }
     }
+
+    fn operand_from_scalar(&self, scalar: Scalar, ty: Ty<'tcx>) -> Operand<'tcx> {
+        Operand::Constant(Box::new(
+            Constant {
+                span: DUMMY_SP,
```
Comment (eddyb) [inline src/librustc_mir/transform/const_prop.rs]: Can you keep the span of, say, the enclosing `Statement`?

File: src/librustc_mir/transform/const_prop.rs [hunk 6]
```diff
@@ -497,6 +499,46 @@ impl<'a, 'mir, 'tcx> ConstPropagator<'a, 'mir, 'tcx> {
...
+            match im {
+                interpret::Immediate::Scalar(ScalarMaybeUndef::Scalar(scalar)) => {
+                    *rval = Rvalue::Use(self.operand_from_scalar(scalar, value.layout.ty));
+                },
+                Immediate::ScalarPair(
+                    ScalarMaybeUndef::Scalar(one),
+                    ScalarMaybeUndef::Scalar(two)
+                ) => {
+                    let ty = &value.layout.ty.sty;
+                    if let ty::Tuple(substs) = ty {
```
Comment (eddyb) [inline src/librustc_mir/transform/const_prop.rs]: I don't see why this is needed, you should be able to create a `ty::Const` from any constant.

**Review by eddyb — APPROVED** [review]

**Comment by wesleywiser** [comment]
Pushed fixes

**Comment by wesleywiser** [comment]
> warning: spurious network error (2 tries remaining): failed to get 200 response from `https://static.crates.io/crates/png/png-0.8.0.crate`, got 504
warning: spurious network error (1 tries remaining): failed to get 200 response from `https://static.crates.io/crates/png/png-0.8.0.crate`, got 504
error: failed to download from `https://crates.io/api/v1/crates/png/0.8.0/download`
Caused by:
  failed to get 200 response from `https://static.crates.io/crates/png/png-0.8.0.crate`, got 504
thread 'main' panicked at 'tests failed for https://github.com/servo/webrender', src\tools\cargotest\main.rs:88:9
note: Run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

@bors retry

**Pull request merged**